### PR TITLE
Stable marriage problem implementation in Elvish

### PR DESCRIPTION
**Pull request by andrewandreii**
This is an algorithm I came up with so it needs to be checked.

### CODE REVIEW

File: contents/stable_marriage_problem/code/elvish/stable_marriage.elv [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    exit = 1
+    for female $fem-names {
+      # choose men who have $female as their first choice
+      proposals = [(take-if [elem]{ $proposal-cond $elem $female } $men-names)]
+
+      # if we found proposals
+      if (not (empty $proposals)) {
+        exit = 0
+      } else {
+        continue
```
Comment (Amaras) [inline contents/stable_marriage_problem/code/elvish/stable_marriage.elv:100]: Okay, that one took me a while to understand, but it makes sense. You should say that if we didn't find any proposal, the algorithm has finished.

File: contents/stable_marriage_problem/code/elvish/stable_marriage.elv [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      }
+
+      # eliminate those who can't marry because she's already married
+      if (==s (kind-of $fem-pref[$female]) string) {
+        for elem $proposals {
+          men-pref[$elem] = $men-pref[$elem][1..]
+        }
```
Comment (Amaras) [inline contents/stable_marriage_problem/code/elvish/stable_marriage.elv:107]: I don't understand that loop, could you explain it better? No one is married until the end of the algorithm. So... what is that doing?

**Review by Amaras — COMMENTED** [review]
I don't really understand the language, but the **Gale-Shapley** (and not Gale-Shapely) algorithm looks correct, though there are a few weird points I'd like you to clarify. I haven't run it though, so testing is not yet done.

Thanks for your contribution!

File: contents/stable_marriage_problem/code/elvish/stable_marriage.elv (hunk 1 above)
Comment (andrewandreii) [reply contents/stable_marriage_problem/code/elvish/stable_marriage.elv:99]: I have changed the condition to be `is-empty $proposals` (also `empty` renamed to `is-empty`) and modified the comment with more explanation of what the algorithm is doing.

File: contents/stable_marriage_problem/code/elvish/stable_marriage.elv (hunk 2 above)
Comment (andrewandreii) [reply contents/stable_marriage_problem/code/elvish/stable_marriage.elv:108]: When $female has a pair, $fem-pref[$female] is equal to the male's name, i.e., it's no longer a map but a string. So we need to remove $female from all the males who are in $proposals.

**Comment by ntindle** [comment]
[lang: elvish]

**Pull request closed without merging**